lib/klaviyo/client.rb

```diff
@@ -10,57 +10,76 @@ class Client
     HTTP_PUT = 'put'
 
     ALL = 'all'
+    METRIC = 'metric'
+    METRICS = 'metrics'
     TIMELINE = 'timeline'
 
     DEFAULT_COUNT = 100
     DEFAULT_PAGE = 0
     DEFAULT_SORT_DESC = 'desc'
 
+    CONTENT_JSON = 'application/json'
+    CONTENT_URL_FORM = 'application/x-www-form-urlencoded'
+
     private
 
-    def self.request(method, path, kwargs = {})
+    def self.request(method, path, content_type, **kwargs)
       check_private_api_key_exists()
       url = "#{BASE_API_URL}/#{path}"
       connection = Faraday.new(
         url: url,
         headers: {
-          'Content-Type' => 'application/json'
+          'Content-Type' => content_type
       })
+      if content_type == CONTENT_JSON
+        kwargs[:body] = kwargs[:body].to_json
+      end
       response = connection.send(method) do |req|
-        req.body = kwargs[:body].to_json || nil
+        req.body = kwargs[:body] || nil
       end
     end
 
-    def self.public_request(method, path, kwargs = {})
+    def self.public_request(method, path, **kwargs)
       check_public_api_key_exists()
       params = build_params(kwargs)
       url = "#{BASE_API_URL}/#{path}?#{params}"
       res = Faraday.get(url).body
     end
 
-    def self.v1_request(method, path, kwargs = {})
-      defaults = {:page => nil,
-                  :count => nil,
-                  :since => nil,
-                  :sort => nil}
-      params = defaults.merge(kwargs)
-      query_params = encode_params(params)
-      full_url = "#{V1_API}/#{path}?api_key=#{Klaviyo.private_api_key}#{query_params}"
-      request(method, full_url)
+    def self.v1_request(method, path, content_type: CONTENT_JSON, **kwargs)
+      if content_type == CONTENT_URL_FORM
+        data = {
+          :body => {
+            :api_key => Klaviyo.private_api_key
+          }
+        }
+        data[:body] = data[:body].merge(kwargs[:params])
+        full_url = "#{V1_API}/#{path}"
+        request(method, full_url, content_type, data)
+      else
+        defaults = {:page => nil,
+                    :count => nil,
+                    :since => nil,
+                    :sort => nil}
+        params = defaults.merge(kwargs)
+        query_params = encode_params(params)
+        full_url = "#{V1_API}/#{path}?api_key=#{Klaviyo.private_api_key}#{query_params}"
+        request(method, full_url, content_type)
+      end
     end
 
-    def self.v2_request(method, path, kwargs = {})
+    def self.v2_request(method, path, **kwargs)
       path = "#{V2_API}/#{path}"
       key = {
         "api_key": "#{Klaviyo.private_api_key}"
       }
       data = {}
       data[:body] = key.merge(kwargs)
-      request(method, path, data)
+      request(method, path, CONTENT_JSON, data)
     end
 
     def self.build_params(params)
-      "data=#{Base64.encode64(JSON.generate(params)).gsub(/\n/,'')}"
+      "data=#{CGI.escape(Base64.encode64(JSON.generate(params)).gsub(/\n/, ''))}"
```

Comment:
this line's getting a little beefy, assuming it was pretty thoroughly tested?

Reply:
yes, it was actually this way before I started making updates to the SDK.  I removed it because I didn't think it was necessary but now there have been a few ZD tickets that made it clear that it's needed after all